grype/pkg/package_test.go

```diff
@@ -0,0 +1,229 @@
+						ArtifactID: "pom-artifact-id-info",
+						Version:    "pom-version-info",
+						Extra: map[string]string{
+							"extra-key": "extra-value",
+						},
+					},
+				},
+			},
+			metadata: JavaMetadata{
+				VirtualPath:   "virtual-path-info",
+				PomArtifactID: "pom-artifact-id-info",
+				PomGroupID:    "pom-group-id-info",
+				ManifestName:  "main-section-name-info",
+			},
+		},
+		// below cases we assert that there is no mapped metadata
+		{
+			name: "apk-metadata",
+			syftPkg: syftPkg.Package{
+				MetadataType: syftPkg.ApkMetadataType,
+				Metadata: syftPkg.ApkMetadata{
+					Package:          "a",
+					OriginPackage:    "a",
+					Maintainer:       "a",
+					Version:          "a",
+					License:          "a",
+					Architecture:     "a",
+					URL:              "a",
+					Description:      "a",
+					Size:             1,
+					InstalledSize:    1,
+					PullDependencies: "a",
+					PullChecksum:     "a",
+					GitCommitOfAport: "a",
+				},
+			},
+			metadata: nil,
+		},
+		{
+			name: "npm-metadata",
+			syftPkg: syftPkg.Package{
+				MetadataType: syftPkg.NpmPackageJSONMetadataType,
+				Metadata: syftPkg.NpmPackageJSONMetadata{
+					Author:      "a",
+					Homepage:    "a",
+					Description: "a",
+					URL:         "a",
+				},
+			},
+			metadata: nil,
+		},
+		{
+			name: "python-metadata",
+			syftPkg: syftPkg.Package{
+				MetadataType: syftPkg.PythonPackageMetadataType,
+				Metadata: syftPkg.PythonPackageMetadata{
+					Name:                 "a",
+					Version:              "a",
+					License:              "a",
+					Author:               "a",
+					AuthorEmail:          "a",
+					Platform:             "a",
+					SitePackagesRootPath: "a",
+				},
+			},
+			metadata: nil,
+		},
+		{
+			name: "gem-metadata",
+			syftPkg: syftPkg.Package{
+				MetadataType: syftPkg.GemMetadataType,
+				Metadata: syftPkg.GemMetadata{
+					Name:     "a",
+					Version:  "a",
+					Homepage: "a",
+				},
+			},
+			metadata: nil,
+		},
+		{
+			name: "kb-metadata",
+			syftPkg: syftPkg.Package{
+				MetadataType: syftPkg.KbPackageMetadataType,
+				Metadata: syftPkg.KbPackageMetadata{
+					ProductID: "a",
+					Kb:        "a",
+				},
+			},
+			metadata: nil,
+		},
+		{
+			name: "rust-metadata",
+			syftPkg: syftPkg.Package{
+				MetadataType: syftPkg.RustCargoPackageMetadataType,
+				Metadata: syftPkg.CargoPackageMetadata{
+					Name:     "a",
+					Version:  "a",
+					Source:   "a",
+					Checksum: "a",
+				},
+			},
+			metadata: nil,
+		},
+	}
+
+	// capture each observed metadata type, we should see all of them relate to what syft provides by the end of testing
+	observedMetadataTypes := set.NewStringSet()
+	expectedMetadataTypes := set.NewStringSet()
+	for _, ty := range syftPkg.AllMetadataTypes {
+		expectedMetadataTypes.Add(string(ty))
+	}
+
+	// run all of our cases
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			observedMetadataTypes.Add(string(test.syftPkg.MetadataType))
```

Comment:
I think this might just be a case of the Mondays....... but I'm not clear on what's being tested with `observedMetadataTypes`

Reply:
the comments within the test comment on what is being tested with those values:
```
// capture each observed metadata type, we should see all of them relate to what syft provides by the end of testing
...
// did we see all possible metadata types? if not, then there is an uncovered case and this test should error out
```

Reply:
@wagoodman Okay thanks, this is helping...

So we don't want Syft to have more metadata types than are tested here in Grype.

I'm trying to explore in my head why I took so long to understand that. (Of course, part of it is just me 😄 )

I think one factor is that we were testing two things at once. It could be a future improvement is to extract that assertion logic into its own function whose name states intent (e.g. `assertGrypeCaseForEverySyftMetadataType` or something). It also sounds like "expected/observed" maps to "Syft/Grype" — a change to `syftMetadataTypes` and `grypeMetadataTypes` (or similar) might help readers understand why we're concerned with tracking and diffing these sets.

Also, if I'm understanding correctly, there shouldn't be a case where Grype is testing more metadata types than are coming from Syft, since Grype wouldn't be able to reference symbols from Syft that Syft doesn't have in the currently used version. If we had messed up in Syft and left symbols around that Syft wasn't using, I guess this could open up the possibility for this case.

One fun fact, possibly useful / possibly not, there's an `assert.ElementsMatch` function that can compare slices in such a way that ignores order, which could help us avoid building sets in some cases.